python/ray/serve/application.py

```diff
@@ -0,0 +1,376 @@
+            raise TypeError(f"Got {type(deployment)}. Expected deployment.")
+        elif deployment.name in self._deployments:
+            raise ValueError(
+                f'Deployment with name "{deployment.name}" already '
+                "exists in this application. To overwrite this "
+                "deployment, use attribute or index notation. "
+                "For example:\n\napp.deployment_name = "
+                "new_deployment"
+            )
+
+        self._deployments[deployment.name] = deployment
+
+    def deploy(self, blocking: bool = True):
+        """
+        Deploys all the deployments contained in this Application to the Ray
+        cluster atomically. Deployments in this Application can carry handles
+        to one another.
+
+        Args:
+            blocking (bool): If True, this function only returns after the
+                deployment is finished. If False, this function returns
+                immediately after requesting the deployment.
+        """
+
+        if len(self._deployments) == 0:
+            return
+
+        parameter_group = []
+
+        for deployment in self._deployments.values():
+            if not isinstance(deployment, Deployment):
+                raise TypeError(
+                    f"deploy_group only accepts Deployments, but got unexpected "
+                    f"type {type(deployment)}."
+                )
+
+            deployment_parameters = {
+                "name": deployment._name,
+                "func_or_class": deployment._func_or_class,
+                "init_args": deployment.init_args,
+                "init_kwargs": deployment.init_kwargs,
+                "ray_actor_options": deployment._ray_actor_options,
+                "config": deployment._config,
+                "version": deployment._version,
+                "prev_version": deployment._prev_version,
+                "route_prefix": deployment.route_prefix,
+                "url": deployment.url,
+            }
+
+            parameter_group.append(deployment_parameters)
+
+        return _get_global_client().deploy_group(parameter_group, _blocking=blocking)
+
+    @staticmethod
+    def get_statuses() -> Dict[str, DeploymentStatus]:
+        """
+        Gets the currents statuses of all deployments running on the Ray cluster.
+        NOTE: these deployments may not match the deployments stored in this
+        Application since other Applications and deployments may have been
+        deployed to the cluster.
+
+        Returns:
+            Dict[str, DeploymentStatus]: This dictionary maps the running
+                deployment's name to a DeploymentStatus object containing its
+                status and a message explaining the status.
+        """
+
+        return _get_global_client().get_deployment_statuses()
+
+    def run(
+        self,
+        runtime_env_updates=None,
+        cluster_address="auto",
+        dashboard_address="http://localhost:8265",
```

Comment:
I don't think any of these should be included:

- We can expect users to call ray.init before this function or else set `RAY_ADDRESS`.
- Let's leave runtime_env_updates as an implementation detail of the CLI for now.
- Don't need `dashboard_address` because the dynamic uploads are already handled if users do `ray.init(working_dir=".")`.

Reply:
Got it– I made the necessary changes.